$OpenBSD: patch-util_wmmenugen_parse_xdg_c,v 1.1 2017/08/25 11:01:17 dcoppa Exp $

commit e037ae3684928a2fbf4a3994562a322f5d3b0c71
Author: Christophe CURIS <christophe.curis@free.fr>
Date:   Sun Jul 2 21:01:49 2017

util: fix parsing of XDG menus with multiple groups

The specification for XDG menu files allows having more than one group and
assumes no constraint on their order. The original code assumed that once
the proper group marker was found, everything after was part of it, causing
misinterpretation of some data, like taking the wrong launch command.

This patch stops the parsing of the menu XDG file when a new group is
found, hence implementing the expected behaviour.

Reported-by: Andreas Metzler <ametzler@bebt.de>

Index: util/wmmenugen_parse_xdg.c
--- util/wmmenugen_parse_xdg.c.orig
+++ util/wmmenugen_parse_xdg.c
@@ -132,6 +132,11 @@ void parse_xdg(const char *file, cb_add_menu_entry *ad
 			/* start processing group */
 			memset(buf, 0, sizeof(buf));
 			continue;
+		} else if (p[0] == '[') {
+			/* If we find a new group and the previous group was the main one,
+			 * we stop all further processing
+			 */
+			if (InGroup) break;
 		}
 
 		if (!InGroup) {
